File: app/src/main/java/io/treehouses/remote/InitialActivity.java
package io.treehouses.remote;

import android.Manifest;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.widget.Toolbar;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import com.google.android.material.navigation.NavigationView;

import io.treehouses.remote.Fragments.AboutFragment;
import io.treehouses.remote.Fragments.HomeFragment;
import io.treehouses.remote.Fragments.NetworkFragment;
import io.treehouses.remote.Fragments.ServicesFragment;
import io.treehouses.remote.Fragments.SettingsFragment;
import io.treehouses.remote.Fragments.StatusFragment;
import io.treehouses.remote.Fragments.SystemFragment;
import io.treehouses.remote.Fragments.TerminalFragment;
import io.treehouses.remote.Fragments.TunnelFragment;
import io.treehouses.remote.Network.BluetoothChatService;
import io.treehouses.remote.bases.PermissionActivity;
import io.treehouses.remote.callback.HomeInteractListener;
import io.treehouses.remote.callback.NotificationCallback;
import io.treehouses.remote.utils.GPSService;
import io.treehouses.remote.utils.LogUtils;


public class InitialActivity extends PermissionActivity
        implements NavigationView.OnNavigationItemSelectedListener, HomeInteractListener, NotificationCallback {

    private static InitialActivity instance = null;
    private Boolean validBluetoothConnection = false;
    int REQUEST_COARSE_LOCATION = 99;
    private static BluetoothChatService mChatService = null;
    private String mConnectedDeviceName = null;
    private NavigationView navigationView;
    DrawerLayout drawer;
    private String TAG = "InitialActivity";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        instance = this;
        setContentView(R.layout.activity_initial2);
        requestPermission();
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        drawer = findViewById(R.id.drawer_layout);

        checkLocationPermission();

        if (mChatService == (null)) {
            Log.e(TAG, "mChatService Status: NULL");
            mChatService = new BluetoothChatService(mHandler);
        } else {
            Log.e(TAG, "mChatService Status: " + mChatService.getState());
            mChatService.updateHandler(mHandler);
        }

        checkStatusNow();

        openCallFragment(new HomeFragment());

        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);
        navigationView.setItemIconTintList(null);
//        navigationView.addHeaderView(getResources().getLayout(R.layout.navigation_view_header));
        new GPSService(this);
    }

    public static InitialActivity getInstance() {
        return instance;
    }

    @Override
    public void onBackPressed() {
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            Fragment f = (getSupportFragmentManager()).findFragmentById(R.id.fragment_container);
            if(f instanceof HomeFragment) finish();
            super.onBackPressed();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.initial, menu);
        return true;
    }

    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();
        checkStatusNow();
        if (validBluetoothConnection) {
            onNavigationItemClicked(id);
        } else {
            if (id == R.id.menu_about) {
                openCallFragment((new AboutFragment()));
            } else if (id == R.id.menu_home) {
                openCallFragment(new HomeFragment());
            } else {
                showAlertDialog();
            }
        }

        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    private void onNavigationItemClicked(int id) {
        if (id == R.id.menu_home) {
            openCallFragment(new HomeFragment());
        } else if (id == R.id.menu_network) {
            openCallFragment((new NetworkFragment()));
        } else if (id == R.id.menu_system) {
            openCallFragment(new SystemFragment());
        } else if (id == R.id.menu_terminal) {
            openCallFragment(new TerminalFragment());
        } else {
            checkMore(id);
        }
    }

    private void checkMore(int id) {
        if (id == R.id.menu_services) {
            openCallFragment(new ServicesFragment());
        } else if (id == R.id.menu_tunnel) {
            openCallFragment(new TunnelFragment());
        } else if (id == R.id.menu_about) {
            openCallFragment((new AboutFragment()));
        } else if (id == R.id.menu_status) {
            openCallFragment(new StatusFragment());
        } else {
            openCallFragment(new HomeFragment());
        }
    }

    @Override
    public void openCallFragment(androidx.fragment.app.Fragment newfragment) {
        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
        fragmentTransaction.replace(R.id.fragment_container, newfragment);
        fragmentTransaction.addToBackStack("");
        fragmentTransaction.commit();
//        menuItem.setChecked(true);
        setTitle("Treehouses Remote");
//        drawer.closeDrawers();

    }
//
    @Override
    public void setNotification(Boolean b) {
        if (b) navigationView.getMenu().getItem(7).setIcon(R.drawable.status_notification);
        else navigationView.getMenu().getItem(7).setIcon(R.drawable.status);
    }

    protected void checkLocationPermission() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION)
                != PackageManager.PERMISSION_GRANTED) {

            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.ACCESS_COARSE_LOCATION},
                    REQUEST_COARSE_LOCATION);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           String permissions[], int[] grantResults) {
        switch (requestCode) {
            case 99: {
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                } else {
                    //TODO re-request
                }
                break;
            }
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        mChatService.updateHandler(mHandler);
    }


    @Override
    public void setChatService(BluetoothChatService chatService) {
        mChatService = chatService;
        mChatService.updateHandler(mHandler);
        checkStatusNow();
    }

    @Override
    public BluetoothChatService getChatService() {
        return mChatService;
    }

    private void checkStatusNow() {
        if (mChatService.getState() == Constants.STATE_CONNECTED) {
            LogUtils.mConnect();
            validBluetoothConnection = true;
        } else if (mChatService.getState() == Constants.STATE_NONE) {
            LogUtils.mOffline();
            validBluetoothConnection = false;
        } else {
            LogUtils.mIdle();
            validBluetoothConnection = false;
        }
        Log.e("BOOLEAN", "" + validBluetoothConnection);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_settings) {
            openCallFragment(new SettingsFragment());
        }
        return super.onOptionsItemSelected(item);
    }

    private AlertDialog showAlertDialog() {
        return new AlertDialog.Builder(InitialActivity.this)
                .setTitle("ALERT:")
                .setMessage("Connect to raspberry pi via bluetooth in the HOME PAGE first before accessing this feature")
                .setIcon(R.drawable.bluetooth)
                .setNegativeButton("OK", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.cancel();
                    }
                })
                .show();
    }

    /**
     * Sends a message.
     *
     * @param message A string of text to send.
     */
    @Override
    public void sendMessage(String message) {
        // Check that we're actually connected before trying anything
        LogUtils.log(message);
        if (mChatService.getState() != Constants.STATE_CONNECTED) {
            Toast.makeText(InitialActivity.this, R.string.not_connected, Toast.LENGTH_SHORT).show();
            LogUtils.mIdle();
            return;
        }

        // Check that there's actually something to send
        if (message.length() > 0) {
            // Get the message bytes and tell the BluetoothChatService to write
            byte[] send = message.getBytes();
            mChatService.write(send);

            // Reset out string buffer to zero and clear the edit text field
//            mOutStringBuffer.setLength(0);
        }
    }

    /**
     * The Handler that gets information back from the BluetoothChatService
     */
    public final Handler mHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
//            FragmentActivity activity = getActivity();
            //InitialActivity activity = InitialActivity.this;
            switch (msg.what) {
                case Constants.MESSAGE_DEVICE_NAME:
                    // save the connected device's name
                    mConnectedDeviceName = msg.getData().getString(Constants.DEVICE_NAME);
                    if (mConnectedDeviceName != "" || mConnectedDeviceName != null) {
                        Log.e("DEVICE", "" + mConnectedDeviceName);
                        checkStatusNow();
//                        Toast.makeText(InitialActivity.this, "Connected to "+mConnectedDeviceName, Toast.LENGTH_LONG).show();
                    }
//                    if (null != activity) {
//                        Toast.makeText(activity, "Connected to "
//                                + mConnectedDeviceName, Toast.LENGTH_SHORT).show();
//                    }
                    break;
            }
        }
    };


}

File: app/src/main/kotlin/io/treehouses/remote/InitialActivity.kt
package io.treehouses.remote

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.preference.PreferenceManager
import io.treehouses.remote.MainApplication
import io.treehouses.remote.callback.BackPressReceiver
import io.treehouses.remote.databinding.ActivityInitial2Binding
import io.treehouses.remote.fragments.CommunityFragment
import io.treehouses.remote.fragments.DiscoverFragment
import io.treehouses.remote.fragments.SettingsFragment
import io.treehouses.remote.fragments.ShowBluetoothFileFragment
import io.treehouses.remote.fragments.dialogfragments.FeedbackDialogFragment
import io.treehouses.remote.ui.home.HomeFragment
import io.treehouses.remote.utils.DialogUtils
import io.treehouses.remote.utils.GPSService

class InitialActivity : BaseInitialActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bind = ActivityInitial2Binding.inflate(layoutInflater)
        instance = this
        setContentView(bind.root)
        requestPermission()
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        currentTitle = "Home"
        setUpDrawer()
        title = "Home"

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED &&
            ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED &&
            ContextCompat.checkSelfPermission(this, Manifest.permission.FOREGROUND_SERVICE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            startGPSService()
        }

        mChatService.updateHandler(mHandler)
        checkStatusNow()
        openCallFragment(HomeFragment())

        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                handleCustomOnBackPressed()
            }
        })
    }

    private fun startGPSService() {
        val intent = Intent(this, GPSService::class.java)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(intent)
        } else {
            startService(intent)
        }
    }

    override fun onResume() {
        super.onResume()
        resetMenuIcon()
    }

    private fun setUpDrawer() {
        mActionBarDrawerToggle = object : ActionBarDrawerToggle(this, bind.drawerLayout, findViewById(R.id.toolbar), R.string.navigation_drawer_open, R.string.navigation_drawer_close) {
            override fun onDrawerOpened(drawerView: View) {
                (this@InitialActivity.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager).hideSoftInputFromWindow(window.decorView.windowToken, 0)
            }
        }
        bind.drawerLayout.addDrawerListener(mActionBarDrawerToggle)
        mActionBarDrawerToggle.syncState()
        bind.navView.setNavigationItemSelectedListener(this)
    }

    fun handleCustomOnBackPressed() {
        if (bind.drawerLayout.isDrawerOpen(GravityCompat.START)) bind.drawerLayout.closeDrawer(GravityCompat.START)
        else {
            val f = supportFragmentManager.findFragmentById(R.id.fragment_container)
            if (f is HomeFragment) finishAffinity()
            else if (f is SettingsFragment || f is CommunityFragment || f is DiscoverFragment || f is ShowBluetoothFileFragment) {
                (supportFragmentManager).popBackStack()
                title = currentTitle
            }
            if (f is BackPressReceiver) f.onBackPressed()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.initial, menu)
        return true
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            REQUEST_LOCATION_PERMISSION_FOR_COMMUNITY -> {
                if ((grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED)) {
                    preferences?.edit()?.putBoolean("send_log", true)?.apply()
                    goToCommunity()
                } else {
                    Toast.makeText(this, "Permission denied. Cannot proceed to community features.", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    override fun onPrepareOptionsMenu(menu: Menu?): Boolean {
        checkStatusNow()
        for (x in 1 until bind.navView.menu.size() - 2) {
            val item = bind.navView.menu.getItem(x)
            item.isEnabled = validBluetoothConnection
        }
        return super.onPrepareOptionsMenu(menu)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        mChatService.updateHandler(mHandler)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> {
                openCallFragment(SettingsFragment())
                title = getString(R.string.action_settings)
            }
            R.id.action_feedback -> {
                FeedbackDialogFragment().show(supportFragmentManager.beginTransaction(), "feedbackDialogFragment")
            }
            R.id.action_community -> {
                if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                    dataSharing()
                } else {
                    preferences?.edit()?.putBoolean("send_log", true)?.apply()
                    goToCommunity()
                }
            }
        }
        return super.onOptionsItemSelected(item)
    }

    fun dataSharing() {
        preferences = PreferenceManager.getDefaultSharedPreferences(this)
        val v = layoutInflater.inflate(R.layout.alert_log_map, null)
        if (preferences?.getBoolean("send_log", false) == false) {
            val builder = DialogUtils.createAlertDialog(this@InitialActivity,
                "Sharing is Caring",
                "To enable the community map, Treehouses needs to collect some data and your approximate location. Would you like to enable data sharing?", v)
                .setCancelable(false)
            DialogUtils.createAdvancedDialog(builder, Pair("Enable Data Sharing", "Cancel"), {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION), REQUEST_LOCATION_PERMISSION_FOR_COMMUNITY)
            }, {MainApplication.showLogDialog = false })
        } else {
            goToCommunity()
        }
    }

    private fun goToCommunity() {
        openCallFragment(CommunityFragment())
        title = getString(R.string.action_community)
    }

    private fun showLocationPermissionDisclosureForCommunity() {
        AlertDialog.Builder(this)
            .setTitle("Location & GPS Usage")
            .setMessage("This app collects location data to create a community map and understand user distribution. " +
                    "This helps us improve our services. To continue, please enable GPS.")
            .setPositiveButton("Ok") { _, _ ->
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION), REQUEST_LOCATION_PERMISSION_FOR_COMMUNITY)
            }
                .setNegativeButton("No", null).show()
    }

    fun changeAppBar() {
        mActionBarDrawerToggle = ActionBarDrawerToggle(this, bind.drawerLayout, findViewById(R.id.toolbar), 0, 0)
        mActionBarDrawerToggle.toolbarNavigationClickListener = View.OnClickListener {
            //reset to burger icon
            supportFragmentManager.popBackStack()
            resetMenuIcon()
        }
        //add back button
        bind.drawerLayout.setDrawerListener(mActionBarDrawerToggle)
        mActionBarDrawerToggle.syncState()
        supportActionBar?.setDisplayHomeAsUpEnabled(false)
        mActionBarDrawerToggle.isDrawerIndicatorEnabled = false
        bind.drawerLayout.setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    fun resetMenuIcon() {
        mActionBarDrawerToggle.isDrawerIndicatorEnabled = true
        bind.drawerLayout.setDrawerLockMode(DrawerLayout.LOCK_MODE_UNLOCKED)
    }

    companion object {
        private const val REQUEST_LOCATION_PERMISSION_FOR_COMMUNITY = 2
    }
}
